### PR TITLE
softmax/logsoftmax optimization

### PR DESCRIPTION
**Pull request by scxiao**
optimization of softmax and logsoftmax operators

### CODE REVIEW

File: src/targets/gpu/include/migraphx/gpu/device/reduce_opers.hpp [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+                                    std::size_t item_num,
+                                    std::size_t max_index)
+{
+    while(true)
+    {
+        auto stride = (item_num + 1) / 2;
+        auto size   = item_num / 2;
+        for(std::size_t i = thr_idx; i < size; i += block_size)
+        {
+            // data_ptr[i] = ::max(to_hip_type(data_ptr[i]), to_hip_type(data_ptr[i + stride]));
```
Comment (kahmed10) [inline src/targets/gpu/include/migraphx/gpu/device/reduce_opers.hpp]: extra commented code

File: src/targets/gpu/include/migraphx/gpu/device/reduce_opers.hpp [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        __syncthreads();
+        item_num = stride;
+
+        if(item_num == 1)
+            break;
+    }
+
+    if(thr_idx == 0)
+    {
+        // data_ptr[max_index] =
```
Comment (kahmed10) [inline src/targets/gpu/include/migraphx/gpu/device/reduce_opers.hpp]: extra commented code

File: src/targets/gpu/device/logsoftmax.cpp [hunk 3]
```diff
@@ -11,53 +12,87 @@ inline namespace MIGRAPHX_INLINE_NS {
...
+                    lds_data[thr_idx] = input[data_idx];
+                }
+                __syncthreads();
 
-            for(std::size_t j = 0; j < num_in_batch; ++j)
-            {
-                data_idx[axis]   = j;
-                output[data_idx] = input[data_idx] - batch_max;
+                auto item_num = (remaining_item_num > block_size) ? block_size : remaining_item_num;
+                // reduce_max(lds_data, block_size, thr_idx, item_num, max_block_size);
```
Comment (kahmed10) [inline src/targets/gpu/device/logsoftmax.cpp]: extra commented code

File: src/targets/gpu/device/logsoftmax.cpp [hunk 4]
```diff
@@ -11,53 +12,87 @@ inline namespace MIGRAPHX_INLINE_NS {
...
+                {
+                    data_idx[axis]    = i;
+                    lds_data[thr_idx] = input[data_idx] - batch_max;
+                    lds_data[thr_idx] = ::exp(to_hip_type(lds_data[thr_idx]));
+                }
+
+                __syncthreads();
+
+                auto item_num = (remaining_item_num > block_size) ? block_size : remaining_item_num;
+                // reduce_sum(lds_data, block_size, thr_idx, item_num, max_block_size);
```
Comment (kahmed10) [inline src/targets/gpu/device/logsoftmax.cpp]: extra commented code

File: src/targets/gpu/device/logsoftmax.cpp (hunk 3 above)
Comment (scxiao) [reply src/targets/gpu/device/logsoftmax.cpp]: fixed

File: src/targets/gpu/device/logsoftmax.cpp (hunk 4 above)
Comment (scxiao) [reply src/targets/gpu/device/logsoftmax.cpp]: fixed

**Pull request closed without merging**